### ISSUE TITLE
Recompilation of brainvisa in casa-distro py3 uses py2

### ISSUE THREAD

**Issue by Hboni** [brainvisa/casa-distro#49]
I have a casa-distro with brainvisa compiled in python 3 (by adding the option in the bv_maker : 
```
cmake_options = -DPYTHON_EXECUTABLE=/usr/bin/python3
```

But when I try to make a bv_maker on this existing distro, it runs without problems, but when I try to launch brainvisa, an error message appears:

![Capture du 2020-04-30 11-36-38](https://user-images.githubusercontent.com/27290853/80696349-91fb7900-8ad7-11ea-9904-36ea5ff244b8.png)

One way to fix it is to to delete the soma folder, and make a new bv_maker, so it download and compile from scratch without problem in python3.

**Comment by ylep** [comment]
Did you switch an existing build tree from Python 2 to Python 3? This is is not supported, if you want to do that you must delete your build tree and recompile entirely. Otherwise the compiled Python libraries will not be rebuilt, hence the version mismatch.

**Comment by Hboni** [comment]
No I didn't switch, I changed anything in the configuration before building, I just wanted to update my actual python3 build.

**Comment by ylep** [comment]
So, if I understand correctly:

1. You have a working Python 3 build;
2. You run `casa_distro bv_maker`, which completes successfully;
3. The new build gives a *version mismatch* error.

Now, which folder exactly did you delete to fix the issue? Was it the entire build tree, or a specific sub-directory?

**Comment by denisri** [comment]
I have sometimes seen similar situations, where the python version have changed unexpectedly in the CMakeCache (for an unknown reason, cmake sometimes resets some variables in a way I don't understand) and things have been compiled partly with python2 and partly with python3. So yes you have to remove/rebuild at least parts with python bindings. Perhaps you can do in the build tree:
```find . -name 'sip*.cpp' -exec rm "{}" \;```
and restart bv_maker build

**Comment by Hboni** [comment]
To be sure I used to delete the entire build tree, but if you have a solution to not have to recompile everything.
I tried what you proposed @denisri, and I made a complete bv_maker. It ends with the same problem of python version incompatibility. Maybe I must only do a `bv_maker build`?

**Comment by denisri** [comment]
Which python version is recorded in CMakeCache.txt ?
you can check using ldd on python/soma/aims/aimssip.so and see if it's linked against python2 or python3.

**Comment by denisri** [comment]
If it's Ok for aims, there are also a number of other python binding modules, so a better check should be done, but also you can try using python3/ipython3 to import soma.aims and see if it works.

**Comment by sapetnioc** [comment]
If there are wrong values in CMake cache, it may be necessary to also
remove the CMakeCache.txt file so that it is recreated by configure
(hopefully with correct values).

**Comment by denisri** [comment]
Once a wrong info gets in the CMakeCache.txt, it will remain here since the job of the cache is indeed to keep values and reuse them instead of detecting them again. This is why once the cache gets corrupted/wrong, it should be completely removed.

**Comment by Hboni** [comment]
Yeah, but the first build in python3 works, creating a good CMakeCache, but trying to recompile modify this CMakeCache to use python 2 instead of python 3. For now, I'm just going to delete the CMakeCache before bv_maker, but it seems there is a problem.

**Comment by denisri** [comment]
Do you have a reproducible way of causing the problem ? How do you recompile before it goes wrong ?

**Comment by Hboni** [comment]
I just run a bv_maker normally, I couldn't reproduce the bug if I tried to recompile directly after a good compilation, it seems that it needs some changes. 
But I didn't try a lot of conditions. I'll make more try during the afternoon, to try to be sure to reproduce the problem.
